### ISSUE TITLE
DAOImpl#deleteById(Collection<T>) ignores Converter

### ISSUE THREAD

**Issue by micmeyer** [jOOQ/jOOQ#12931]
### Expected behavior
When calling `DAOImpl#deleteById(Collection<T> ids)`, with more than one element in the `ids` list,  the org.jooq.Converter for the type T should be used.

### Actual behavior
The converter for the type T is ignored. Instead the constructor `java.util.UUID(byte[])` is used for converting the custom type to a UUID.
This leads to broken UUIDs. eg. the UUID 21720a1e-9049-41b2-9522-f969615d473c is transformed to 32313732-3061-3165-2d39-3034392d3431

The feature **works as expected** if the "ids" list only contains **a single value**.

### Steps to reproduce the problem

- Create a custom type `FooId`
- Create a custom converter that has a UUID as "fromType" (`FooIdConverter`)
- Create a list that contains multiple FooIds (**Must be more than one**)
- Call `DAOImpl#deleteById(Collection<T> ids)`

```
public class FooId {

    private final UUID id;

    public FooId(UUID id) {
        this.id = id;
    }

    public UUID getId() {
        return id;
    }

    public static FooId fromString(String id) {
        return new FooId(UUID.fromString(id));
    }
}

public class FooIdConverter implements Converter<UUID, FooId> {

    @Override
    public FooId from(UUID databaseObject) {
        return databaseObject == null ? null : new FooId(databaseObject);
    }

    @Override
    public UUID to(FooId userObject) {
        return userObject == null ? null : userObject.getId();
    }

    @Override
    public Class<UUID> fromType() {
        return UUID.class;
    }

    @Override
    public Class<FooId> toType() {
        return FooId.class;
    }
}
```

### Versions

- jOOQ: 3.15.4 (Problem can also be reproduced with 3.15.6)
- Java: Zulu13.44+13-CA
- Database (include vendor): PostgreSQL 13
- OS: MacOS
- JDBC Driver (include name if inofficial driver): 42.3.1

**Comment by lukaseder** [comment]
Thanks for your report. How is your converter connected to the DAO?

**Comment by micmeyer** [comment]
We use the following forced type:
```
'userType'  : 'FooId',
'converter' : 'FooIdConverter',
'includeExpression': '.*\\.FOO_ID',
'includeTypes'     : '.*'
```

**Comment by lukaseder** [comment]
OK, and what does your table look like? (`CREATE TABLE ...`)

**Comment by micmeyer** [comment]
Create Table statement:

```
CREATE TABLE foo
(
    id    UUID PRIMARY KEY,
    value VARCHAR(255) NOT NULL
);

INSERT INTO foo(id, value)
VALUES ('b15eaa22-c0eb-4837-bd64-a81266b058e6', 'a value');
INSERT INTO foo(id, value)
VALUES ('3555aaec-29ea-4733-b845-498dec7b2ada', 'an other value');
```

Test code:
```
List<FooId> fooIds = Lists.newArrayList(FooId.fromString("b15eaa22-c0eb-4837-bd64-a81266b058e6"), FooId.fromString("3555aaec-29ea-4733-b845-498dec7b2ada"));
dao().deleteById(fooIds);
```

**Comment by lukaseder** [comment]
The root cause is that `ConvertedDataType` overrides `AbstractDataType.convert(Object)`, but not the overloads accepting `convert(Object...)` or `convert(Collection<?>)`, so it's not strictly related to DAOs, though it does manifest here.

This simpler test also fails, for the same reason:

```java
@Test
public void testConvertedDataTypeConvert() {
    DataType<Integer> t = VARCHAR.asConvertedDataType(
        Converter.ofNullable(String.class, Integer.class, Integer::valueOf, Object::toString)
    );

    assertEquals(1, t.convert("1"));
    assertEquals(asList(1, 2), t.convert("1", "2"));
}
```

The exception being:

```
java.lang.AssertionError: expected:<[1, 2]> but was:<[Ljava.lang.Integer;@3e2059ae>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at org.jooq.test.AbstractTest.assertEquals(AbstractTest.java:194)
	at org.jooq.test.ConverterTest.testConvertedDataTypeConvert(ConverterTest.java:447)
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.3 (https://github.com/jOOQ/jOOQ/issues/12935), and 3.15.7 (https://github.com/jOOQ/jOOQ/issues/12936). Thanks again for the report!
